plat/common/include/x86/paging.h

```diff
@@ -0,0 +1,326 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Authors: Marc Rittinghaus <marc.rittinghaus@kit.edu>
+ *
+ * Copyright (c) 2021, Karlsruhe Institute of Technology (KIT).
+ *                     All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef __PLAT_CMN_ARCH_PAGING_H__
+#error Do not include this header directly
+#endif
+
+#include <uk/arch/limits.h>
+#include <uk/essentials.h>
+#include <uk/plat/paging.h>
+#include <uk/fallocbuddy.h>
+#include <uk/print.h>
+
+#include <errno.h>
+
+/* We do not support 32-bit virtual address spaces as they do not provide
+ * enough space for the direct mapping of page tables. In this case, we would
+ * have to use temporary mappings or recursive page table mapping
+ */
+#if PT_LEVELS < 4
+#error "Unsupported number of page table levels"
+#endif
+
+/*
+ * We expect the physical memory to be mapped 1:1 into the following area in
+ * every virtual address space. This allows us to easily translate from virtual
+ * to physical page table addresses and vice versa. We also access the metadata
+ * for the frame allocators through this area.
+ */
+#define DIRECTMAP_AREA_START	0xffffff8000000000 /* -512 GiB */
+#define DIRECTMAP_AREA_END	0xffffffffffffffff
+#define DIRECTMAP_AREA_SIZE	(DIRECTMAP_AREA_END - DIRECTMAP_AREA_START + 1)
+
+static inline __vaddr_t
+x86_directmap_paddr_to_vaddr(__paddr_t paddr)
+{
+	UK_ASSERT(paddr < DIRECTMAP_AREA_SIZE);
```

Comment:
Should these be conditional to `CONFIG_UK_DEBUG`?

Reply:
Good question. But in fact, this would mean that every `UK_ASSERT` in any platform code would need to be enclosed, which is dislike. I think we are better off with moving the assert to a public header and find a transparent way to have `ukdebug` hook into it.

Reply:
Exactly, that's the way 👍🏼 And with the opportunity remove the conditionals from the arch headers in this PR :)

Reply:
I think for this header `UK_ASSERT` is fine to have here as is. So far, we handled dependencies of platform-internal headers with the platforms that use them. So, each platform using this `paging.h` should express a dependency to `lib/ukdebug`. Obviously, this gets a bit messy to retrieve all dependencies for shared platform code in the future but we were thinking of a re-organization of the platform code anyways since a while. We should come up with a scheme that - ideally - automatically creates platform dependencies.... I also had another discussion with Marc in PR #227. An idea might be to move runtime and compiletime assertions to `/include/uk`. We mainly just need to remove the hard-dependency to `ukdebug` printing whenever an assertion fails.